### PR TITLE
[16.0][ADD] edi_utm_oca

### PR DESCRIPTION
**Pull request by chaule97**
-  this module `edi_utm_oca` automatically sets the EDI source on models that have a source_id field when they are created.

### CODE REVIEW

File: edi_sale_utm_oca/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    "website": "https://github.com/OCA/edi-framework",
+    "author": "Camptocamp, Odoo Community Association (OCA)",
+    "depends": ["utm", "edi_sale_oca"],
+    "data": [
+        "data/utm_source.xml",
+        "views/edi_exchange_record_views.xml",
+    ],
+    "installable": True,
+    "auto_install": True,
+}
```
Comment (QuocDuong1306) [inline edi_sale_utm_oca/__manifest__.py]: Please remove `"auto_install": True`

Add `"maintainers": ["simahawk"],`

Change to `Allows to define an utm source on an exchange type and automatically set it on the sale order when it is created.`

File: edi_sale_utm_oca/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# Copyright 2024 Camptocamp
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+{
+    "name": "EDI Sale utm",
```
Comment (QuocDuong1306) [inline edi_sale_utm_oca/__manifest__.py]:
```suggestion
    "name": "EDI Sale UTM",
```

File: edi_sale_utm_oca/models/edi_exchange_type.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+from odoo import api, fields, models
+
+
+class EDIExchangeType(models.Model):
+    _inherit = "edi.exchange.type"
+
+    edi_source_id = fields.Many2one(
+        "utm.source", default=lambda self: self._default_edi_source_id()
+    )
```
Comment (QuocDuong1306) [inline edi_sale_utm_oca/models/edi_exchange_type.py]: Maybe need a `help`

File: edi_sale_utm_oca/models/sale_order.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+
+from odoo import api, models
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    def _set_source_id(self, vals):
```
Comment (QuocDuong1306) [inline edi_sale_utm_oca/models/sale_order.py]:
```suggestion
    def _edi_set_source_id(self, vals):
```

File: edi_sale_utm_oca/models/sale_order.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+        if vals.get("origin_exchange_type_id"):
+            exchange_type = self.env["edi.exchange.type"].browse(
+                vals["origin_exchange_type_id"]
+            )
+            if exchange_type and exchange_type.edi_source_id:
+                vals["source_id"] = exchange_type.edi_source_id.id
```
Comment (QuocDuong1306) [inline edi_sale_utm_oca/models/sale_order.py]: Just set if `source_id` not in `vals`

File: edi_sale_utm_oca/models/sale_order.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        for vals in vals_list:
+            self._set_source_id(vals)
+
```
Comment (QuocDuong1306) [inline edi_sale_utm_oca/models/sale_order.py]:
```suggestion
```

File: edi_sale_utm_oca/tests/test_sale_order.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+                "name": "",
+                "origin_exchange_type_id": self.exchange_type_in.id,
+                "partner_id": self.partner.id,
+            }
+        )
+
+        self.assertEqual(
+            sale_order.source_id.id,
+            self.env.ref("edi_sale_utm_oca.utm_source_edi").id,
+        )
```
Comment (QuocDuong1306) [inline edi_sale_utm_oca/tests/test_sale_order.py]: reuse the test in `edi_sale_oca` instead

You can have a look at `edi_sale_ubl_oca` (v14, `edi/pull/759`) to get the idea to test

**Review by QuocDuong1306 — CHANGES_REQUESTED** [review]
Also maybe missing `CONFIGURE.rst`

**Review by QuocDuong1306 — COMMENTED** [review]
Sorry, I think the good approach is creating 2 modules instead: `edi_utm_oca` and `edi_sale_utm_oca` (benefits: to reuse this feature on stock, product, purchase,.. in the future if needed)

You can reference `edi_state_oca` and `edi_sale_ubl_oca` to have an idea

discuss with me, If you have difficulty when researching

File: edi_sale_utm_oca/models/sale_order.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+            if all(conditions):
+                vals["edi_source_id"] = exchange_type.edi_source_id.id
+
+        return vals
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        for vals in vals_list:
+            self._set_source_edi(vals)
+        return super().create(vals_list)
```
Comment (QuocDuong1306) [inline edi_sale_utm_oca/models/sale_order.py]: Why you don't use `_edi_set_source` on `edi_utm_oca` 

And please put this one in `def _edi_set_source` of `edi_utm_oca`

```
if vals.get("origin_exchange_record_id"):
            exchange_record = self.env["edi.exchange.record"].browse(
                vals["origin_exchange_record_id"]
            )

            exchange_type = exchange_record.type_id

            conditions = [
                exchange_type,
                exchange_type.edi_source_id,
                "edi_source_id" not in vals,
            ]

            if all(conditions):
                vals["edi_source_id"] = exchange_type.edi_source_id.id

File: edi_sale_utm_oca/readme/DESCRIPTION.rst [hunk 9]
```diff
@@ -0,0 +1 @@
+This module allows to set a utm source on an exchange type and sets it automatically when the order is created.
```
Comment (QuocDuong1306) [inline edi_sale_utm_oca/readme/DESCRIPTION.rst]:
```suggestion
This module allows to set a utm source on an exchange type and sets it automatically when the sale order is created.
```

File: edi_sale_utm_oca/tests/test_sale_order.py [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+            "origin_exchange_record_id": exchange_record.id,
+            "partner_id": self.partner.id,
+        }
+
+        sale_order = self._create_sale_order(**vals)
+
+        self.assertEqual(
+            sale_order.edi_source_id.id,
+            self.env.ref("edi_utm_oca.utm_source_edi").id,
+        )
```
Comment (QuocDuong1306) [inline edi_sale_utm_oca/tests/test_sale_order.py]: You can test the same with [this test](https://github.com/OCA/edi-framework/blob/16.0/edi_sale_oca/tests/test_order.py) (can import to reuse if needed)

**Review by QuocDuong1306 — CHANGES_REQUESTED** [review]

File: edi_sale_utm_oca/models/sale_order.py (hunk 8 above)
Comment (QuocDuong1306) [reply edi_sale_utm_oca/models/sale_order.py]: sorry, please ignore the block, just need to re-use `_edi_update_source`

File: edi_sale_utm_oca/__manifest__.py [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+    "summary": """
+    Automatically set edi source on the sale order when it is created.
+    """,
+    "version": "16.0.1.0.0",
+    "development_status": "Alpha",
+    "license": "LGPL-3",
+    "website": "https://github.com/OCA/edi-framework",
+    "author": "Camptocamp, Odoo Community Association (OCA)",
+    "maintainers": ["simahawk"],
+    "depends": ["sale_management", "edi_utm_oca"],
```
Comment (QuocDuong1306) [inline edi_sale_utm_oca/__manifest__.py]:
```suggestion
    "depends": ["edi_sale_oca", "edi_utm_oca"],
```
We need the `origin_exchange_record_id` on Sale Order

File: edi_sale_utm_oca/tests/test_sale_order.py [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+        cls._setup_order(
+            origin_exchange_record_id=cls.exc_record_in.id,
+        )
+
+    def test_source_edi(self):
+        order = self.sale
+        self.assertEqual(
+            order.edi_source_id.id,
+            self.env.ref("edi_utm_oca.utm_source_edi").id,
+        )
```
Comment (QuocDuong1306) [inline edi_sale_utm_oca/tests/test_sale_order.py]:
```suggestion
        self.assertEqual(
            order.edi_source_id,
            self.env.ref("edi_utm_oca.utm_source_edi"),
        )
```

**Review by QuocDuong1306 — CHANGES_REQUESTED** [review]
Could you create a view to put `edi_source_id` into [this group](https://github.com/OCA/edi-framework/blob/d2dd6121f838d3a788b3a318776b7e50dccc37f3/edi_sale_oca/views/sale_order.xml#L12)

**Review by QuocDuong1306 — APPROVED** [review]
LG,

Thank a Chau

File: edi_sale_utm_oca/models/sale_order.py [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+
+from odoo import api, models
+
+
+class SaleOrder(models.Model):
+    _name = "sale.order"
+    _inherit = [
+        "sale.order",
+    ]
```
Comment (QuocDuong1306) [inline edi_sale_utm_oca/models/sale_order.py]:
```suggestion
    _inherit = "sale.order"
```

File: edi_sale_utm_oca/readme/DESCRIPTION.rst [hunk 14]
```diff
@@ -0,0 +1 @@
+This module allows to set a utm source on an exchange type and sets it automatically when the sale order is created.
```
Comment (QuocDuong1306) [inline edi_sale_utm_oca/readme/DESCRIPTION.rst]:
```suggestion
This module allows to set a utm source on an exchange type and sets it automatically when the sale order is created through this exchange type.
```

**Review by QuocDuong1306 — APPROVED** [review]
Pre-approving

**Review by QuocDuong1306 — APPROVED** [review]

**Comment by chaule97** [comment]
Hi @QuocDuong1306 , can you review my PR again

File: edi_utm_oca/readme/DESCRIPTION.rst [hunk 15]
```diff
@@ -0,0 +1 @@
+This module automatically sets the EDI source on models that have a source_id field when they are created.
```
Comment (QuocDuong1306) [inline edi_utm_oca/readme/DESCRIPTION.rst]:
```suggestion
This module automatically assigns the EDI source to records created through the EDI mechanism.
```

**Review by QuocDuong1306 — APPROVED** [review]
Pre-approving

**Pull request merged**